lib/mlir/Pipes/MLIRContainer.cpp

```diff
@@ -0,0 +1,270 @@
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include <optional>
+
+#include "mlir/Bytecode/BytecodeReader.h"
+#include "mlir/Bytecode/BytecodeWriter.h"
+#include "mlir/Dialect/LLVMIR/LLVMDialect.h"
+#include "mlir/IR/FunctionInterfaces.h"
+#include "mlir/IR/IRMapping.h"
+#include "mlir/IR/SymbolTable.h"
+#include "mlir/Parser/Parser.h"
+
+#include "revng/Pipeline/RegisterContainerFactory.h"
+
+#include "revng-c/Pipes/Kinds.h"
+#include "revng-c/mlir/Pipes/MLIRContainer.h"
+
+#include "MLIRLLVMHelpers.h"
+
+using namespace revng;
+using namespace revng::pipes;
+using namespace revng::mlir_llvm;
+
+static constexpr auto &TheKind = kinds::MLIRLLVMFunctionKind;
+
+using ContextPtr = std::unique_ptr<mlir::MLIRContext>;
+using OwningModuleRef = mlir::OwningOpRef<mlir::ModuleOp>;
+
+static OwningModuleRef cloneModuleInto(mlir::ModuleOp SourceModule,
+                                       mlir::MLIRContext &DestinationContext) {
```

Comment:
Please document this function: why does it serialize the module (to have the source module in a different context)? What's the downside of this approach (pay serialization)? What would be the ideal solution (cross-context cloning)?

The function is rather trivial but the reasons why we do things this way, tradeoffs and alternatives should be documented.